src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21706,6 +21706,484 @@ void M(string? s)
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""missing"")] public string? M(string? p) => throw null!;
+
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1).ToString(); // 1
+        c.M(p2).ToString(); // 2
+    }
+}";
+            // Warn on misused attribute? https://github.com/dotnet/roslyn/issues/36073
+            var comp2 = CreateNullableCompilation(new[] { source, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(
+                // (9,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p1).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p1)").WithLocation(9, 9),
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p2).ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p2)").WithLocation(10, 9)
+                );
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_NullName()
+        {
+            var source =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(null)] public string? M(string? p) => throw null!; // 1
+
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1).ToString(); // 2
+        c.M(p2).ToString(); // 3
+    }
+}";
+            var comp2 = CreateNullableCompilation(new[] { source, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(
+                // (4,31): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //     [return: NotNullIfNotNull(null)] public string? M(string? p) => throw null!; // 1
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(4, 31),
+                // (9,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p1).ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p1)").WithLocation(9, 9),
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p2).ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p2)").WithLocation(10, 9)
+                );
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_ParamsParameter()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p"")] public string? M(params string?[]? p) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string?[]? p1, string?[]? p2, string? p3, string? p4)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1).ToString();
+        c.M(p2).ToString(); // 1
+
+        _ = p3 ?? throw null!;
+        c.M(p3).ToString(); // 2
```

Comment:
I don't understand these warnings. The params array is definitely non-null in both of these cases.

Reply:
Thanks. Indeed

---
In reply to: [303562190](https://github.com/dotnet/roslyn/pull/37201#discussion_r303562190) [](ancestors = 303562190)

Reply:
Could you add an actual `null` argument as one of the tests? So we can see it fail?